src/vm/collection.rs

```diff
@@ -4,12 +4,25 @@ use crate::util::opaque_pointer::*;
 use crate::vm::VMBinding;
 use crate::{scheduler::*, Mutator};
 
+use super::scanning::QueuingTracerFactory;
+
 /// Thread context for the spawned GC thread.  It is used by spawn_gc_thread.
 pub enum GCThreadContext<VM: VMBinding> {
     Controller(Box<GCController<VM>>),
     Worker(Box<GCWorker<VM>>),
 }
 
+/// Information related to weak reference processing.  Used by `Collection::process_weak_refs`.
+pub struct ProcessWeakRefsContext {
+    /// `true` if `process_weak_refs` is called during the forwarding phase in MarkCompact.
+    /// Always `false` if the GC is not MarkCompact.
+    pub forwarding: bool,
```

Comment:
You could just separate `process_weak_refs` and `forward_weak_refs`. Based on your [binding code](https://github.com/wks/mmtk-openjdk/blob/c962cab7f609ecf585360e2d48305e59e8770389/mmtk/src/weak_processor/mod.rs#L59), the implementation of process and forward are different as well.

The argument from the point of view of MMTk core is that for forwarding, we should not discover new weak refs, and we do not need a boolean return value. So it is a bit confusing to merge them into one method/type/etc. This also affects `VMProcessWeakRefs`.

Reply:
Done.